### PR TITLE
Implement Delete OP

### PR DESCRIPTION
**Pull request by sumeetgajjar**

### CODE REVIEW

File: crdt_lib/include/list.h [hunk 1]
```diff
@@ -40,18 +40,27 @@ namespace nupad::crdt {
 
         void applyInsertOperation(Operation &operation) {
             auto *insertOp = dynamic_cast<InsertOperation<T> *>(&operation);
-            CHECK_NE(insertOp, nullptr) << "Cannot cast operation to InsertOperation: " << operation;
```
Comment (sumeetgajjar) [inline crdt_lib/include/list.h:43]: Change this to CHECK_NE(op, nullptr) to avoid compilation errors

File: crdt_lib/include/clock.h [hunk 2]
```diff
@@ -5,23 +5,32 @@
...
 namespace nupad::clock {
     class VectorClock {
-        static bool initialized_;
-        static PeerId myPeerId_;
-        static ClockState clockState_;
+        bool initialized_ = false;
+        PeerId myPeerId_;
+        Tick tick_;
     public:
-        static void init(PeerId myPeerId);
```
Comment (sumeetgajjar) [inline crdt_lib/include/clock.h:16]: move constructor definition to cpp file

File: crdt_lib/include/clock.h [hunk 3]
```diff
@@ -5,23 +5,32 @@
...
+        explicit VectorClock(PeerId myPeerId) {
+          myPeerId_ = std::move(myPeerId);
```
Comment (sumeetgajjar) [inline crdt_lib/include/clock.h]: use initializer list instead

Refer Node class constructor under DoublyLinkedList for initializer list syntax

File: crdt_lib/test/vector_clock_test.cpp [hunk 4]
```diff
@@ -8,56 +8,35 @@
...
     expectedClockState = {{myPeerId_, 1}};
-    ASSERT_EQ(nupad::clock::VectorClock::getState(), expectedClockState);
+    ASSERT_EQ(clock_->getState(), expectedClockState);
 
-    nupad::clock::VectorClock::tick();
+    clock_->tick();
     expectedClockState = {{myPeerId_, 2}};
-    ASSERT_EQ(nupad::clock::VectorClock::getState(), expectedClockState);
-
-    const nupad::PeerId otherPeerId("random");
```
Comment (sumeetgajjar) [inline crdt_lib/test/vector_clock_test.cpp:38]: why is this assertion deleted?

File: crdt_lib/include/clock.h [hunk 5]
```diff
@@ -5,23 +5,32 @@
...
 
+#include <utility>
+
 namespace nupad::clock {
     class VectorClock {
-        static bool initialized_;
-        static PeerId myPeerId_;
-        static ClockState clockState_;
+        bool initialized_ = false;
+        PeerId myPeerId_;
```
Comment (sumeetgajjar) [inline crdt_lib/include/clock.h]: make this const

File: crdt_lib/include/clock.h [hunk 6]
```diff
@@ -5,23 +5,32 @@
...
+        }
 
-        static Tick tick(Tick tick = 1);
+        Tick tick();
 
-        static void update(const PeerId &peerId, Tick tick);
+        void update(Tick tick);
 
-        static Tick getTick(const PeerId &peerId = myPeerId_);
+        Tick getTick();
```
Comment (sumeetgajjar) [inline crdt_lib/include/clock.h]: mark const

File: crdt_lib/include/clock.h [hunk 7]
```diff
@@ -5,23 +5,32 @@
...
 
-        static ClockState getState();
+        ClockState getState();
```
Comment (sumeetgajjar) [inline crdt_lib/include/clock.h]: mark const

File: crdt_lib/src/clock.cpp [hunk 8]
```diff
@@ -4,45 +4,28 @@
...
-        myPeerId_ = std::move(myPeerId);
-        CHECK_STRNE(myPeerId_.c_str(), "") << "myPeerId cannot be empty";
-        clockState_[myPeerId_] = 0;
-        initialized_ = true;
-        LOG(INFO) << "VectorClock initialized";
-    }
-
-    Tick VectorClock::tick(const Tick tick) {
+    Tick VectorClock::tick() {
         CHECK(initialized_) << "VectorClock not initialized";
```
Comment (sumeetgajjar) [inline crdt_lib/src/clock.cpp]: remove all intialized check

File: crdt_lib/include/clock.h [hunk 9]
```diff
@@ -5,23 +5,32 @@
...
+#include <utility>
+
 namespace nupad::clock {
     class VectorClock {
-        static bool initialized_;
-        static PeerId myPeerId_;
-        static ClockState clockState_;
+        bool initialized_ = false;
+        PeerId myPeerId_;
+        Tick tick_;
```
Comment (sumeetgajjar) [inline crdt_lib/include/clock.h]: why not use map to store ticks of all peers?

File: crdt_lib/include/list.h (hunk 1 above)
Comment (madhurjain2095) [reply crdt_lib/include/list.h:43]: CHECK_NOTNULL works

File: crdt_lib/test/crdt_list_test.cpp [hunk 10]
```diff
@@ -2,10 +2,31 @@
 // Created by sumeet on 11/23/20.
 //
 
+#include "context.h"
 #include "operation.h"
 #include "test_utils.h"
 #include <gtest/gtest.h>
 
+class ContextElementIdTest : public ::testing::Test {
```
Comment (sumeetgajjar) [inline crdt_lib/test/crdt_list_test.cpp]: Nit, rename this to ContextTestSuite

Comment (sumeetgajjar) [inline crdt_lib/test/crdt_list_test.cpp]: Feel free to move this to a separate file

File: crdt_lib/test/crdt_list_test.cpp [hunk 11]
```diff
@@ -2,10 +2,31 @@
 // Created by sumeet on 11/23/20.
```
Comment (sumeetgajjar) [inline crdt_lib/test/crdt_list_test.cpp]: should we rename this file to something else? e.g. crdt_lib_test

File: crdt_lib/test/list_test.cpp [hunk 12]
```diff
@@ -0,0 +1,78 @@
+//
+// Created by madhurjain on 11/25/20.
+//
+
+#include "list.h"
+#include "gtest/gtest.h"
+
+class ListTest : public ::testing::Test {
```
Comment (sumeetgajjar) [inline crdt_lib/test/list_test.cpp]: Nit, rename this to ListTestSuite

File: crdt_lib/test/vector_clock_test.cpp [hunk 13]
```diff
@@ -8,56 +8,54 @@
 
 class VectorClockTest : public ::testing::Test {
```
Comment (sumeetgajjar) [inline crdt_lib/test/vector_clock_test.cpp]: nit, rename this to VectorClockTestSuite

**Review by sumeetgajjar — COMMENTED** [review]
LGTM

**Review by madhurjain2095 — APPROVED** [review]

**Pull request merged**